{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "outside-dispatch",
   "metadata": {},
   "source": [
    "# Specification\n",
    "\n",
    "- This is a command Line program using standard parameters, STDIN and STDOUT redirection\n",
    "- Program name: missingMotif.py\n",
    "- Input/Output: missingMotif.py < [input file] > [output file]\n",
    "\n",
    "## options:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "humanitarian-think",
   "metadata": {},
   "source": [
    "- --minMotif  minimum motif size to evaluate (int)\n",
    "- --maxMotif maximum motif size to evaluate (int)\n",
    "- --cutoff  Z-score cutoff (float)\n",
    "\n",
    "Ex: missingMotif.py --minMotif=3 --maxMotif=8 --cutoff=-4.0 < [input file] > [output file]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "reliable-directive",
   "metadata": {},
   "source": [
    "# Background\n",
    "Many prokaryotic genomes encode Restriction-Modification systems (RM) that act to protect the cell from invading DNA by cutting at specific sites (frequently short 4-6 base reverse complement palindromes). They also protect host DNA from being cut by modifying any sites within the host DNA that would be sites for the cutting enzyme. It is not unusual to find that these enzymes are adjacent to each other in the host genome. (see: https://en.wikipedia.org/wiki/Restriction_modification_system).\n",
    "\n",
    "Meselson and Stahl (1958) showed that replication of the bacterial dsDNA genome happens in a semi-conservative fashion (see: https://en.wikipedia.org/wiki/Prokaryotic_DNA_replication). The resulting circular chromosome(s) would then be transiently hemi-methylated because the new daughter strand may not been have exposed to the protective methylation of the RM system before its cutting activity arrives. This creates a race between RM cutting and protecting activities may expose the cell to strand breaks.\n",
    "\n",
    "Is a selective advantage incurred for the cell by allowing mutations to accumulate at these sites? If so, then would we expect the host genome to become depleted of RM sites over time? Can we use this depletion process as a signature for the existence of the specific RM systems present in the genome? \n",
    "\n",
    "How can we find sequences that are missing in a genome?\n",
    "\n",
    "Note: \"Stealth\" has recently been accepted for publication, and is now publicly available at: https://git.ucsc.edu/dbernick/stealth\n",
    "\n",
    "\n",
    "\n",
    "# Assignment\n",
    "\n",
    "Write a proper style, commandLine BME205 solution that reads a fasta file and ranks motifs based on how statistically underrepresented the specific motif is. We need only consider sequence motifs that are up to about 8 in length ( this is a program option). In order to use the equation that we developed in class, the minimum size should be at least 3 ( this is an option also). We will find that there are quite a few of these, so we need to specify a statistical cutoff. For this assignment, we will use Z-scores. The scores we are looking for will be negative since we are looking for underrepresented sequences (this cutoff is a program option too). Finally, we need to sort the output by decreasing motif size and then increasing (most negative to least negative) z-score within motif size groups and print it out.\n",
    "\n",
    "Remember that sequences that we get from DNA sequencing are from only one strand of DNA. This means that when we see a AAGGTT, this implies that AACCTT (the reverse complement) is present on the other strand. We should count these sequences as equivalent. \n",
    "\n",
    "Your report should be sorted by Zscore within motif-Size, such that the longest motifs are at the beginning of the report and, within motifs of a given size, print the most significant motifs first. Print the k-mer pairs in alpha order. ( AAA:TTT and not TTT:AAA)\n",
    "\n",
    "The extra credit for this assignment is particularly interesting. If you have time, this will present better results. \n",
    "\n",
    "Here is an example (non extra credit):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "absolute-prompt",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence: reverse     count      Expect Zscore\n",
    "AAAATTTA:TAAATTTT\t835\t1737.66\t-21.66\n",
    "GATTAATA:TATTAATC\t550\t1326.89\t-21.33\n",
    "AATTAATA:TATTAATT\t929\t1839.72\t-21.24\n",
    "AATTAATC:GATTAATT\t977\t1861.79\t-20.51\n",
    "ATAATTAA:TTAATTAT\t1033\t1926.49\t-20.36\n",
    "GAAATTTA:TAAATTTC\t378\t1031.07\t-20.34\n",
    "CCGATCGC:GCGATCGG\t1323\t2284.74\t-20.12\n",
    "GCGATCGA:TCGATCGC\t1221\t2121.78\t-19.56\n",
    "ACGATCGC:GCGATCGT\t1188\t2035.31\t-18.78\n",
    "GTTTAAAA:TTTTAAAC\t519\t1151.17\t-18.63\n",
    "AAATATTA:TAATATTT\t738\t1444.79\t-18.60\n",
    "AAAATTTG:CAAATTTT\t857\t1591.95\t-18.42\n",
    "CATTAATC:GATTAATG\t479\t1037.49\t-17.34\n",
    "ATATATAA:TTATATAT\t271\t736.19\t-17.15\n",
    "ATTTAAAG:CTTTAAAT\t419\t941.16\t-17.02\n",
    "AAATATTG:CAATATTT\t687\t1287.74\t-16.74\n",
    "CAAATTTA:TAAATTTG\t670\t1265.24\t-16.74\n",
    "CTTTAAAC:GTTTAAAG\t468\t995.13\t-16.71"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "medieval-running",
   "metadata": {},
   "source": [
    "# Hints\n",
    "- For **genomes that include multiple fasta records, do NOT concatenate them**. They should be reported as a single genome, though we do not want to create false kmers that would result from concatenation\n",
    "- **Avoid zero divide errors!** Close examination of our null model will show that if we have a zero in the denominator, we will have a zero in the numerator. If you have pre-populated your count dictionary, then you will need to handle dividing by zero.\n",
    "- Only consider kmers that are constructed from the **canonical bases {A, C, G, T}**.\n",
    "- Z-score is calculated as: $\\frac{s-\\mu}{sd}$, where s is the specific count you are converting, and mean and standard deviation are given by µ and sd. Both µ and sd describe the null distribution. In our case, we can treat the expected distribution as a binomial where: $\\mu=np,sd=\\sqrt{np(1-p)}$. n = genome size (approximately), and p = probability of success, given the null distribution.\n",
    "- c(x) defined as the count of x, generally the count of some substring. For example, for kmers of size 4: $K=k_{1}k_{2}k_{3}k_{4}, \n",
    "Pr(K)=\\frac{c(K)}{N},\n",
    "E(K) = \\frac{c(k_{1}k_{2}k_{3})\\cdot c(k_{2}k_{3}k_{4})}{c(k_{2}k_{3})}$\n",
    "(see below for the derivation of our null model) We can then calculate n and p, and derive µ and sd. \n",
    "- To keep your scores lined up, use tabs to separate values in each line. The following format string will be of use: \n",
    "\n",
    "print('{0:8}:{1:8}\\t{2:0d}\\t{3:0.2f}\\t{4:0.2f}'.format(seq, rSeq, count,E,pVal))\n",
    "\n",
    "\n",
    "# Null Model derivation\n",
    "If we assume complete independence of the any of the characters that make up a sequence K, we could approximate the expected count of K, E(K),  as:\n",
    "$\\begin{matrix} \n",
    "K=k_{1}k_{2}k_{3}k_{4}, Pr(k_{i})=\\frac{c(k_{i})}{N} \\\\ \n",
    "E(K)=N * Pr(K) = N * \n",
    "Pr(k_{1})\n",
    "Pr(k_{2})\n",
    "Pr(k_{3})\n",
    "Pr(k_{4}) \n",
    "\\end{matrix} \n",
    "$\n",
    "\n",
    "the above is a markov(0) approximation for the expected count (E) of our sequence K.\n",
    "\n",
    "Rarely is this assumption of independence a good null model, due to many factors including codon bias which drives selection of 3-mers based on underlying amino acid frequencies. We can model this dependence using a markov model, where the probability associated with any symbol is conditioned on the preceding characters. For a kmer of size 4, we can condition our character probabilities using the preceding two bases without requiring use of the count of the specific 4-mer that we are trying to approximate(see note on information content). We need to consider edge conditions in our approximation. In general, we can use a model of order k-2, where k is the size of the kmer as long as we have sufficient data, therefore:\n",
    "\n",
    "$\n",
    "\\begin{matrix} \n",
    "Pr(K)=\\frac{E(K)}{N} = \n",
    "Pr(k_{1})\n",
    "Pr(k_{2}|k_{1})\n",
    "Pr(k_{3}|k_{1}k_{2})\n",
    "Pr(k_{4}|k_{2}k_{3}) \\\\\n",
    "= \\frac{c(k_{1})}{N}\n",
    "\\frac{c(k_{1}k_{2})}{c(k_{1})}\n",
    "\\frac{c(k_{1} k_{2} k_{3})} {c(k_{1} k_{2})}\n",
    "\\frac{c(k_{2} k_{3} k_{4})} {c(k_{2} k_{3})} \\\\\n",
    "= \\frac{1}{N}\n",
    "\\frac{c(k_{1} k_{2} k_{3})c(k_{2} k_{3} k_{4})} {c(k_{2} k_{3})}\n",
    "\\end{matrix}\n",
    "$\n",
    "\n",
    "note: the conditional probability Pr(a | b) is read as the probability of finding sequence a, given sequence b. In this case we treat sequence b as adjacent and left of a, and ask what is the probability of finding sequence a, given that b is immediately to the left of a. This means that we should count the occurrences of string ba, among all strings that start with b (taken as the count of all sequence b).\n",
    "\n",
    "note on information content: What would happen if we were trying to predict the frequencies of 4-mers, and our null model contained the frequency of 4-mers? \n",
    "\n",
    "2021 Note: Make sure to review the module video on this. The derivation is a bit more elaborate.\n",
    "\n",
    "# Notes\n",
    "- I removed all extra credit from this assignment. It's full enough as is.\n",
    "- I considered adding pseudo counts to this assignment but have deferred that to the next (4-Oct)\n",
    "- Added the full derivation of the null model (4-Oct)\n",
    "- clarified the use of E - the expected value, and in our case, an expected motif count. (4-Oct)\n",
    "- clarified the use of the count function c(x) (4-Oct)\n",
    "- clarified the use of counts of strings in conditional probabilities. (4-Oct)\n",
    "- clarified the avoidance of zero divide or the use of kmers that contain non-canonical bases (4-Oct)\n",
    "\n",
    "## 2018 additions\n",
    "\n",
    "- binned the reverse complement sequences. This produces better Evalues and a more concise report. ( 1-Oct)\n",
    "\n",
    "## 2020 addition\n",
    "\n",
    "- its good to think about our kmer counting by asking how many names does this kmer have? For example, the kmer AAAA when viewed from the opposite strand is TTTT. This bit of data has 2 names, so when we are counting we make sure that counts of this kmer by either of its names goes into the same bin. How should we handle a reverse complement palindrome like TATA ? When viewed from either strand, it has the same name.\n",
    "\n",
    "## 2021 addition:\n",
    "\n",
    "- we are using notebooks now to formalize inspections and promote Python learning.\n",
    "\n",
    "## 2023\n",
    "- required submissions to be .py files with command line options, STDIN and STDOUT as input/output\n",
    "- provided example of published solution at: https://git.ucsc.edu/dbernick/stealth\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "korean-brick",
   "metadata": {},
   "source": [
    "## Inspection Intro\n",
    "Provide design level information for your inspection team here.Examples:\n",
    "- How did you count kmers of each size without having to read the data multiple times?\n",
    "\n",
    "    -> for every kmer-size of range(3,8), I iterated through the sequence once and calculated the counts in of the motif of specifc size in a dictionary.  \n",
    "    \n",
    "    \n",
    "- How did you organize the computing of mean and SD for each of the kmer groups ?\n",
    "\n",
    "    -> I created a function that takes the input kmer-counts, probability and genome size. Using these inputs I calculated the mean and SD for every kmer in the stored dictionary. \n",
    "    \n",
    "- How did you compute Z-scores for each kMer in a kMer group?\n",
    "\n",
    "    -> In order to compute Z-scores, I didn't do it for each KMer group, but I iterated through the complete dictinary, and calculted for every motif that is beed identified. \n",
    "    \n",
    "- How did you sort the final report to include most significant Z-scores at the top of each kmer-size group?\n",
    "\n",
    "  -> In order to sort the z-scores withing each kmer-size, i used the python sorted function and lambda function to sort the z-score in the ascending algorithm.)\n",
    "  \n",
    "          sorted_dict = dict(sorted(input_dict.items(), key=lambda item: (-len(item[0]), item[1][-1])))\n",
    "\n",
    "- How did you organize the program ?\n",
    "\n",
    "    -> The program initially calls a main function that reads for command inputs and reads the input file. Further, it calls the class Genome. Using the class object, motifSearch function is called from the main function and then motifscore function is called to calculate all the scores like expected score and z-score. At last, display_score function is called which is also part of the class Genome. This is called to display the scores that is calculated for all the motifs.\n",
    "\n",
    "- What did your inspection team find? Which of the inspection results did you use, how did you resolve them, and which did you decide against? Who was on your inspection team?\n",
    "--> In the motifSearch function, changed to range(0, len(seq)-j +1) from range(3, len(seq)-j +1), cause I was not able to calculate expected score for kmer-size 3. So found motif counts for kmer-size below 3 as well. Inspection team include paul, kerney and ava. \n",
    "\n",
    "# Code that may be useful"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58d16012-a33b-497a-8c1a-ebef9ecdab91",
   "metadata": {},
   "source": [
    "## Inspection Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92462c85-9778-42e3-b6a1-7dac1b1d63cf",
   "metadata": {
    "tags": []
   },
   "source": [
    "In the motifSearch function, changed to range(0, len(seq)-j +1) from range(3, len(seq)-j +1), cause I was not able to calculate expected score for kmer-size 3."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "differential-championship",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "class FastAreader:\n",
    "    def __init__(self, fname=''):\n",
    "        '''contructor: saves attribute fname '''\n",
    "\n",
    "        self.fname = fname\n",
    "        self.fileH = None\n",
    "\n",
    "    def doOpen(self):\n",
    "        if self.fname == '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "\n",
    "    def readFasta(self):\n",
    "\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "\n",
    "        with self.doOpen() as self.fileH:\n",
    "\n",
    "            header = ''\n",
    "            sequence = ''\n",
    "\n",
    "            # skip to first fasta header\n",
    "            line = self.fileH.readline()\n",
    "            while not line.startswith('>'):\n",
    "                line = self.fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            for line in self.fileH:\n",
    "                if line.startswith('>'):\n",
    "                    yield header, sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else:\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "\n",
    "        yield header, sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "czech-irrigation",
   "metadata": {},
   "outputs": [],
   "source": [
    "class CommandLine():\n",
    "    '''\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse, now standard in 2.7 and beyond. \n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help, and an error termination mechanism do-usage_and_die.\n",
    "\n",
    "    attributes:\n",
    "    all arguments received from the commandline using .add_argument will be\n",
    "    avalable within the .args attribute of object instantiated from CommandLine.\n",
    "    For example, if myCommandLine is an object of the class, and requiredbool was\n",
    "    set as an option using add_argument, then myCommandLine.args.requiredbool will\n",
    "    name that option.\n",
    "\n",
    "    '''\n",
    "\n",
    "    def __init__(self, inOpts=None):\n",
    "        '''\n",
    "        CommandLine constructor.\n",
    "        Implements a parser to interpret the command line argv string using argparse.\n",
    "        '''\n",
    "\n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(\n",
    "            description='Program prolog - a brief description of what this thing does',\n",
    "            epilog='Program epilog - some other stuff you feel compelled to say',\n",
    "            add_help=True,  # default is True\n",
    "            prefix_chars='-',\n",
    "            usage='%(prog)s [options] -option1[default] <input >output'\n",
    "        )\n",
    "\n",
    "        self.parser.add_argument('-l', '--minMotif', nargs='?', default=1, action='store',\n",
    "                                 help='min kMer size ')\n",
    "        self.parser.add_argument('-m', '--maxMotif', nargs='?', default=8, action='store',\n",
    "                                 help='max kMer size ')\n",
    "        self.parser.add_argument('-c', '--cutoff', nargs='?', type=float, default=.01, action='store',\n",
    "                                 help='Zscore cutoff')\n",
    "\n",
    "        self.parser.add_argument('-v', '--version', action='version', version='%(prog)s 0.1')\n",
    "        if inOpts is None:\n",
    "            self.args = self.parser.parse_args()\n",
    "        else:\n",
    "            self.args = self.parser.parse_args(inOpts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fuzzy-rendering",
   "metadata": {},
   "outputs": [],
   "source": [
    "class CommandLine():\n",
    "    '''\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse, now standard in 2.7 and beyond. \n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help, and an error termination mechanism do-usage_and_die.\n",
    "\n",
    "    attributes:\n",
    "    all arguments received from the commandline using .add_argument will be\n",
    "    avalable within the .args attribute of object instantiated from CommandLine.\n",
    "    For example, if myCommandLine is an object of the class, and requiredbool was\n",
    "    set as an option using add_argument, then myCommandLine.args.requiredbool will\n",
    "    name that option.\n",
    "    '''\n",
    "\n",
    "    def __init__(self, inOpts=None):\n",
    "        '''\n",
    "        CommandLine constructor.\n",
    "        Implements a parser to interpret the command line argv string using argparse.\n",
    "        '''\n",
    "\n",
    "        import argparse\n",
    "\n",
    "        self.parser = argparse.ArgumentParser(\n",
    "            description='Program prolog - a brief description of what this thing does',\n",
    "            epilog='Program epilog - some other stuff you feel compelled to say',\n",
    "            add_help=True,  # default is True\n",
    "            prefix_chars='-',\n",
    "            usage='%(prog)s [options] -option1[default] <input >output'\n",
    "        )\n",
    "\n",
    "        self.parser.add_argument('-l', '--minMotif', nargs='?', default=1, action='store',\n",
    "                                 help='min kMer size ')\n",
    "        self.parser.add_argument('-m', '--maxMotif', nargs='?', default=8, action='store',\n",
    "                                 help='max kMer size ')\n",
    "        self.parser.add_argument('-c', '--cutoff', nargs='?', type=float, default=.01, action='store',\n",
    "                                 help='Zscore cutoff')\n",
    "        self.parser.add_argument('-v', '--version', action='version', version='%(prog)s 0.1')\n",
    "\n",
    "        if inOpts is None:\n",
    "            self.args = self.parser.parse_args()\n",
    "        else:\n",
    "            self.args = self.parser.parse_args(inOpts)\n",
    "\n",
    "class Genome:\n",
    "    \"\"\"\n",
    "    This class helps in storing different attributes that can be used across several functions across the class. \n",
    "    Functions:\n",
    "        reverseComplement: to get the reverse complement of the input sequence. \n",
    "\n",
    "    \"\"\"\n",
    "    def __init__(self, minK, maxK, cutoff, input_sequence):\n",
    "        \"\"\"Construct the object to count and analyze kMer frequencies\"\"\"\n",
    "        self.min_kmer = int(minK)\n",
    "        self.max_kmer = int(maxK)\n",
    "        self.motif_counts = dict() #dict to store motif_counts\n",
    "        self.genome_sequence = input_sequence\n",
    "        self.cutoff = float(cutoff)\n",
    "        self.genome_size = len(self.genome_sequence)\n",
    "\n",
    "    def reverseComplement(self, input_seq):\n",
    "        \"\"\"\n",
    "        reverse complement of the input sequence. \n",
    "\n",
    "        :returns - returns the reverse complement of the input motif. \n",
    "        \"\"\"\n",
    "\n",
    "        rev_seq_str = input_seq[::-1]\n",
    "        reversse_compliment_seq = ''\n",
    "        for i in rev_seq_str:\n",
    "            if i=='A':\n",
    "                reversse_compliment_seq = reversse_compliment_seq + 'T'\n",
    "            elif i=='T':\n",
    "                reversse_compliment_seq = reversse_compliment_seq + 'A'\n",
    "            elif i=='C':\n",
    "                reversse_compliment_seq = reversse_compliment_seq + \"G\"\n",
    "            else:\n",
    "                reversse_compliment_seq = reversse_compliment_seq + \"C\"\n",
    "        return reversse_compliment_seq\n",
    "\n",
    "    def motifSearch(self):\n",
    "        \"\"\"\n",
    "        Function to find all the k-mers for size 3 to 8. It also calculates the frequency of each kmer in the input sequence. \n",
    "        output: \n",
    "            motif_counts: stored as class object. a dictonary with key is the kmers and value are the counts of the kmers. Saved in class variable self.motifCounts. stores all the motifs and counts in self.motif_counts\n",
    "        \"\"\"\n",
    "        seq = self.genome_sequence\n",
    "        for j in range(1, self.max_kmer+1):\n",
    "            for k in range(0, len(seq) - j+ 1):\n",
    "                motif = seq[k:k + j]\n",
    "                if motif in self.motif_counts.keys(): \n",
    "                    rc = self.reverseComplement(motif)\n",
    "                    self.motif_counts[motif] += 1\n",
    "                    self.motif_counts[rc] = self.motif_counts[motif]\n",
    "                else:\n",
    "                    self.motif_counts[motif] = 1\n",
    "                    rc = self.reverseComplement(motif)\n",
    "                    self.motif_counts[rc] = 1\n",
    "\n",
    "    def calcProbab(self, e):\n",
    "        \"\"\"\n",
    "        Calculates Probability of a kmer appearing in a sequence.\n",
    "\n",
    "        input: expect score of kmer. \n",
    "        return: probability value. \n",
    "        \"\"\"\n",
    "        prob = e/self.genome_size\n",
    "        return prob\n",
    "\n",
    "    def zScore(self, s, n, p):\n",
    "        \"\"\"\n",
    "        This function calculates the z-score for a motif.\n",
    "        input: \n",
    "            s - counts of a motif. \n",
    "            n - genomic size\n",
    "            p - probability. \n",
    "\n",
    "        :return - returns the z-score if standard deviation is not zero, else returns 1\n",
    "        \"\"\"\n",
    "        sd = math.sqrt(n*p*(1-p))\n",
    "        if sd != 0:\n",
    "            return (s - (n * p)) / sd\n",
    "        else:\n",
    "            return 1\n",
    "\n",
    "    def expectedScore(self, motif):\n",
    "        \"\"\"\n",
    "        Calculates the expected score of motif using HMM(2). It is done using the derived equation that consists of three terms.\n",
    "        1st term - Considers the motif excluding the last nucleotide. \n",
    "        2nd term - considers the motif excluding the first nucleotide. \n",
    "        3rd term (denominator) - considers the middle part of the motif, excluding the first and last nucleotide.\n",
    "\n",
    "        return: expect score\n",
    "        The function returns zero if the motif is not found in the motif_counts dictonary. \n",
    "        \"\"\"\n",
    "        if motif in self.motif_counts:\n",
    "            first_part =  motif[:-1]\n",
    "            second_part = motif[1:]\n",
    "            denominator_part = motif[1:-1]\n",
    "\n",
    "            if first_part in self.motif_counts and second_part in self.motif_counts and denominator_part in self.motif_counts:\n",
    "                first_motif_counts = self.motif_counts[first_part]\n",
    "                second_motif_counts = self.motif_counts[second_part]\n",
    "                denominator_motif_counts = self.motif_counts[denominator_part]\n",
    "                est_score = (first_motif_counts * second_motif_counts)/(denominator_motif_counts)\n",
    "                return est_score\n",
    "            else:\n",
    "                return 0\n",
    "\n",
    "    def motifScores(self):\n",
    "        \"\"\"\n",
    "        Calls the function expectedScore(), zScore(), and calcProbab() to calculate expected_score and z-score for every motif that is stored in motif_counts dictionary. \n",
    "        \n",
    "        :return - a dictionary (final_dict) which consists of all the motifs as 'key' and its 'values' are [reverse component, expected score and z-score].\n",
    "        \"\"\"\n",
    "        is_motif_calculated = dict()\n",
    "        final_dict = dict()\n",
    "        for motif, counts in self.motif_counts.items():\n",
    "            rc_motif = self.reverseComplement(motif)\n",
    "            if motif not in is_motif_calculated and rc_motif not in is_motif_calculated:\n",
    "                kmers = [motif, rc_motif]\n",
    "                exp_score = self.expectedScore(motif)\n",
    "                s_count = self.motif_counts[motif]\n",
    "                prob = self.calcProbab(exp_score)\n",
    "                z_score = self.zScore(s_count, self.genome_size, prob)\n",
    "                if z_score <= self.cutoff:\n",
    "                    kmers.sort()\n",
    "                    final_dict[kmers[0]] = [kmers[1], counts, exp_score, z_score]\n",
    "                is_motif_calculated[motif] = True\n",
    "                is_motif_calculated[rc_motif] = True\n",
    "        return final_dict\n",
    "    \n",
    "    def display_scores(self, input_dict):\n",
    "        \"\"\"\n",
    "        Displays all the motifs with size in decending orders and within every motif-size, sorts the z-score in ascending order. \n",
    "\n",
    "\n",
    "        :input- a dictionary that consists of motifs as 'keys' and values as the following [reverse component, counts, expected score, and z-score]\n",
    "        :return- prints the output in required format. \n",
    "        \"\"\"\n",
    "        sorted_dict = dict(sorted(input_dict.items(), key=lambda item: (-len(item[0]), item[1][-1])))\n",
    "\n",
    "        print(\"N = {}\".format(self.genome_size-8))\n",
    "        for key,values in sorted_dict.items():\n",
    "            print(\"{}:{}\\t {}\\t {}\\t {}\".format(key, values[0], values[1], values[2], values[3]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "buried-airline",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [options] -option1[default] <input >output\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f /Users/pratikkatte/Library/Jupyter/runtime/kernel-bb478e9f-e569-4b9c-a639-b27b010218a2.json\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/pratikkatte/miniconda3/envs/mercia/lib/python3.11/site-packages/IPython/core/interactiveshell.py:3534: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "def main(inFile=\"\", options = None):\n",
    "\n",
    "    ''' Setup necessary objects, read data and print the final report.'''\n",
    "    cl = CommandLine() # setup the command line\n",
    "    sourceReader = FastAreader() # setup the Fasta reader Object\n",
    "    sequenceList = []\n",
    "    for head, seq in sourceReader.readFasta():       # reading the fast file. \n",
    "        sequenceList.extend(seq)\n",
    "    \n",
    "    sequenceList = \"\".join(sequenceList) # concatenating all the sequencing into one single string. \n",
    "\n",
    "    thisGenome = Genome(cl.args.minMotif, cl.args.maxMotif, cl.args.cutoff, sequenceList) # setup a Genome object\n",
    "\n",
    "    k = thisGenome.motifSearch()\n",
    "    output = thisGenome.motifScores()\n",
    "    thisGenome.display_scores(output)\n",
    "\n",
    "if __name__ == \"__main__\":  \n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ad9f954-d165-4b37-9926-88a328e114ef",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
