{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "cef8682a-4c88-4269-9569-849f6f5f9766",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found Consencus GAAAAACTTAAAA with score - 8.217968381511612\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import math\n",
    "import random\n",
    "from collections import Counter\n",
    "\n",
    "\n",
    "class FastAreader:\n",
    "    def __init__(self, fname=''):\n",
    "        '''contructor: saves attribute fname '''\n",
    "\n",
    "        self.fname = fname\n",
    "        self.fileH = None\n",
    "\n",
    "    def doOpen(self):\n",
    "        if self.fname == '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "\n",
    "    def readFasta(self):\n",
    "\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "\n",
    "        with self.doOpen() as self.fileH:\n",
    "\n",
    "            header = ''\n",
    "            sequence = ''\n",
    "\n",
    "            # skip to first fasta header\n",
    "            line = self.fileH.readline()\n",
    "            while not line.startswith('>'):\n",
    "                line = self.fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "\n",
    "            for line in self.fileH:\n",
    "                if line.startswith('>'):\n",
    "                    yield header, sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else:\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "        yield header, sequence\n",
    "\n",
    "class CommandLine():\n",
    "    \"\"\"\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse, now standard in 2.7 and beyond.\n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help, and an error termination mechanism do-usage_and_die.\n",
    "\n",
    "    attributes:\n",
    "    all arguments received from the commandline using .add_argument will be\n",
    "    available within the .args attribute of object instantiated from CommandLine.\n",
    "    For example, if myCommandLine is an object of the class, and requiredbool was\n",
    "    set as an option using add_argument, then myCommandLine.args.requiredbool will\n",
    "    name that option.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, inOpts=None):\n",
    "        \"\"\"\n",
    "        CommandLine constructor.\n",
    "        There are three arguments that are passed to the class Consensus():\n",
    "            -i iterations (int)\n",
    "            -k motif length (int)\n",
    "            -p pseudocount (float)\n",
    "        \"\"\"\n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(\n",
    "            description='Program prolog - a brief description of what this thing does',\n",
    "            epilog='Program epilog - some other stuff you feel compelled to say',\n",
    "            add_help=True,  # default is True\n",
    "            prefix_chars='-',\n",
    "            usage='python randomizedMotifSearch.py -i int --maxMotif int --cutoff int < input.fa > output.out'\n",
    "        )\n",
    "        # Be sure to go over the argument information again\n",
    "        self.parser.add_argument('-i', type=int, action='store',\n",
    "                                 help='number of iterations (int)')\n",
    "        self.parser.add_argument('-k', type=int, action='store',\n",
    "                                 help='motif length (int)')\n",
    "        self.parser.add_argument('-p', type=int, action='store',\n",
    "                                 help='pseudocount (float)')\n",
    "\n",
    "        if inOpts is None:\n",
    "            self.args = self.parser.parse_args()\n",
    "        else:\n",
    "            self.args = self.parser.parse_args(inOpts)\n",
    "\n",
    "class Genome:\n",
    "    \"\"\"\n",
    "    This class helps in initializing and storing different attributes that can be used across several functions across the class. \n",
    "    \n",
    "    \"\"\"\n",
    "    def __init__(self, i, k, p, input_sequences):\n",
    "        \"\"\"\n",
    "\n",
    "        \"\"\"\n",
    "        self.iterations = i\n",
    "        self.pseudo_counts = p\n",
    "        self.kmer_size = k\n",
    "        self.input_sequences = input_sequences\n",
    "        self.best_score = 0\n",
    "        self.best_motif = []\n",
    "        self.null_distribution = self.null_model_distribution()\n",
    "        \n",
    "    def newMotif(self, profile):\n",
    "        \"\"\"\n",
    "        Function: To get new set of motifs from the existing profiles.\n",
    "        input: profile of a set of existing selected motifs. \n",
    "        \n",
    "        return: new set of motifs that generated using the counts of bases from the profile. \n",
    "        \"\"\"\n",
    "\n",
    "\n",
    "        new_motif = []\n",
    "        for seq in self.input_sequences:\n",
    "            best_prod_score = 0\n",
    "            best_motif = []\n",
    "\n",
    "            for i in range(0, len(seq)-self.kmer_size+1):\n",
    "                motif = seq[i:i+self.kmer_size]\n",
    "                product_score = 1\n",
    "                for pos, base in enumerate(motif):\n",
    "                    count_from_profile = profile[base][pos]\n",
    "                    product_score = product_score * count_from_profile\n",
    "                if product_score > best_prod_score:\n",
    "                    best_prod_score = product_score\n",
    "                    best_motif = motif\n",
    "                \n",
    "            new_motif.append(best_motif)\n",
    "        \n",
    "        return new_motif\n",
    "\n",
    "    def selectRandomKmer(self, input_sequence_list):\n",
    "        \"\"\"\n",
    "        Function: select random kmers from all the reads in the fast file. \n",
    "        input: input sequence list from the fasta file.\n",
    "        \n",
    "        return: set of random motifs of size k.\n",
    "        \"\"\"\n",
    "        \n",
    "        rand_kmer_array = []\n",
    "        \n",
    "        for sequence in input_sequence_list:\n",
    "            \n",
    "            rand_index = random.randint(0, len(sequence)-self.kmer_size)\n",
    "            \n",
    "            random_kmer = sequence[rand_index:rand_index+self.kmer_size]\n",
    "            rand_kmer_array.append(random_kmer)\n",
    "        return rand_kmer_array\n",
    "    \n",
    "    def null_model_distribution(self):\n",
    "        \"\"\"\n",
    "        Function that returns a dictionary of number of counts of every bases {A,C,G,T} in the complete genome sequence.\n",
    "        \"\"\"\n",
    "        input_seq = self.input_sequences\n",
    "        joined_input_seq = \"\".join(input_seq)\n",
    "        col_counter = dict(Counter(joined_input_seq))\n",
    "\n",
    "        null_distribution = {k: (v+self.pseudo_counts) / (len(joined_input_seq)+ self.pseudo_counts*4) for k, v in col_counter.items()}\n",
    "        \n",
    "        return null_distribution\n",
    "\n",
    "\n",
    "    def make_profile_from_kmers(self, input_motifs):\n",
    "        \"\"\"\n",
    "        The function calculates the distribution of bases for the provided input set of motifs. \n",
    "\n",
    "        input: set of motifs of kmer-size k. \n",
    "        return: a maxtrix of size 4xk. where the four rows are the bases {A,C,G,T} and the columns of size k each having the counts of bases.\n",
    "        \"\"\"\n",
    "        profile_dict = {\n",
    "            'A': [],\n",
    "            'G': [],\n",
    "            'C': [],\n",
    "            'T': []\n",
    "        }\n",
    "\n",
    "        for i in range(self.kmer_size):\n",
    "            col = []\n",
    "            for j in range(len(input_motifs)):\n",
    "                col.append(input_motifs[j][i])\n",
    "            col_counter = dict(Counter(col))\n",
    "            \n",
    "            nset = len(self.input_sequences)\n",
    "            profile_dict['A'].append(col_counter['A']+self.pseudo_counts if 'A' in col_counter else 0+self.pseudo_counts)\n",
    "            profile_dict['C'].append(col_counter['C']+self.pseudo_counts if 'C' in col_counter else 0+self.pseudo_counts)\n",
    "            profile_dict['T'].append(col_counter['T']+self.pseudo_counts if 'T' in col_counter else 0+self.pseudo_counts)\n",
    "            profile_dict['G'].append(col_counter['G']+self.pseudo_counts if 'G' in col_counter else 0+self.pseudo_counts)\n",
    "        return profile_dict\n",
    "    \n",
    "    def calcRelativeEntropy(self, input_motifs):\n",
    "        \"\"\"\n",
    "        The function considers the null model and compares it with the experimental model inorder to calculate the relative entropy.\n",
    "        The function also call the make_profile_from_kmers function to create profile from the provided input set of motifs. \n",
    "\n",
    "        Input: the set of motifs of kmer-size k. \n",
    "\n",
    "        return: the relative entropy score of the provided set of motifs and the profile of the motifs. \n",
    "\n",
    "        \"\"\"\n",
    "        relative_motif_score = 0\n",
    "\n",
    "        profile = self.make_profile_from_kmers(input_motifs)\n",
    "        for col in range(self.kmer_size):\n",
    "            for base in self.null_distribution:\n",
    "                if profile[base][col] != 0: \n",
    "                    pr = profile[base][col]/(len(self.input_sequences) + (4*self.pseudo_counts))\n",
    "                    relative_motif_score += pr * math.log2(pr / self.null_distribution[base])\n",
    "        return relative_motif_score, profile\n",
    "\n",
    "    def randomMotifSearch(self, input_seqs):\n",
    "        \"\"\"\n",
    "        The function iterates for self.iterations times and initially selects random motifs of kmers-size k, and creates profile to select the next set of motifs which will be compared with the \n",
    "        previous set of motifs using relative entropy score. The process is repeated until the best relative entropy score is found. \n",
    "\n",
    "        Input: input sequences of the whole fasta file. \n",
    "        return: prints the consensus and the best relative entropy score of the given input. \n",
    "        \"\"\"\n",
    "        \n",
    "        for i in range(self.iterations):\n",
    "            \n",
    "            random_kmers = self.selectRandomKmer(input_seqs) # initialized kmers randomly with kmer-size -k\n",
    "            best_score_motif, best_profile = self.calcRelativeEntropy(random_kmers) # calculate the score of the randomly initialized kmers. Consider it as the best_score. \n",
    "            selected_motifs = random_kmers\n",
    "            while True:\n",
    "                new_motifs = self.newMotif(best_profile) # select new motifs from the current profile\n",
    "                new_score, new_profile = self.calcRelativeEntropy(new_motifs)\n",
    "                \n",
    "                if new_score > best_score_motif:\n",
    "                    selected_motifs = new_motifs\n",
    "                    best_score_motif = new_score\n",
    "                    best_profile = new_profile\n",
    "\n",
    "                    if best_score_motif > self.best_score:\n",
    "                        self.best_score = best_score_motif\n",
    "                        self.best_motif = selected_motifs\n",
    "                else:\n",
    "                    break\n",
    "\n",
    "        print(\"Found Consencus {} with score - {}\".format(self.printConsencus(self.best_motif), self.best_score))\n",
    "\n",
    "    def printConsencus(self, input_motifs):\n",
    "        \"\"\"\n",
    "        Function: Using the best motif found using randomsearch algorithm, the function returns the consensus from the profile of the motif. \n",
    "\n",
    "        Input: input_motif - is the best motif found from the randmom search algorithm.\n",
    "\n",
    "        return: the consensus motif of kmer-size k. \n",
    "        \"\"\"\n",
    "        motif_set = input_motifs\n",
    "        profile = self.make_profile_from_kmers(motif_set)\n",
    "        concensus = ''\n",
    "        for i in range(self.kmer_size):\n",
    "            max_prob = 0\n",
    "            max_base = 'A'\n",
    "            for base, probs in profile.items():\n",
    "                if probs[i] > max_prob:\n",
    "                    max_prob = probs[i]\n",
    "                    max_base = base\n",
    "            concensus+= max_base\n",
    "        return concensus\n",
    "\n",
    "def main(inFile=\"\", options = None):\n",
    "    ''' Setup necessary objects, read data and print the final report.'''\n",
    "    cl = CommandLine(options) # setup the command line\n",
    "    sourceReader = FastAreader(inFile) # setup the Fasta reader Object\n",
    "    sequenceList = []\n",
    "    for head, seq in sourceReader.readFasta():       # reading the fast file. \n",
    "        sequenceList.append(seq)\n",
    "    \n",
    "    thisGenome = Genome(cl.args.i, cl.args.k, cl.args.p, sequenceList)\n",
    "    thisGenome.randomMotifSearch(sequenceList)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    # main()\n",
    "    inpFile = \"p1860Crisprs\"\n",
    "    main(inpFile, options = [ \"-i 1000\", \"-k 13\", \"-p 1\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c3f1d23-2392-4dee-9546-974e1b4ae86f",
   "metadata": {},
   "source": [
    "# Inspection intro\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f61bd0de-e123-45ff-bcb4-955bfdd59b10",
   "metadata": {},
   "source": [
    "What data structure did you use to hold your sequence motifs (DNA)?\n",
    "\n",
    "-- I used lists to hold the sequence motifs."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "207b4b2f-792d-478e-89b7-0c728d6b87b7",
   "metadata": {},
   "source": [
    "Is your profile structure based on counts or frequencies?\n",
    "\n",
    "-> My profile structure is based on counts. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11be427c-0cb6-4c28-b580-8b9ef85dd925",
   "metadata": {},
   "source": [
    "\n",
    "Where did you implement iterations (mostly why did you choose this)?\n",
    "\n",
    "-> The iterations in the Random Motif Search Algorithm are implemented to repeatedly perform the motif discovery process. The choice of implementing iterations is primarily driven by the need to explore different possibilities and iteratively refine the motifs. Here's why iterations are crucial and why they are chosen in this algorithm\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be375d56-3579-440d-be9e-dbd8bd06acf2",
   "metadata": {
    "tags": []
   },
   "source": [
    "If you implemented Gibbs Sampling, how did you implement recomputing of the profile and motif?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84a30ca8-d36f-47e7-a3b4-2b662cce14a6",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Inspection Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a300c8be-cd4b-48a7-a6cb-608d06b0c0ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PseudoCounts - "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
